modules/tinymce/src/core/main/ts/content/InsertContentImpl.ts

```diff
@@ -58,17 +58,62 @@ const trimBrsFromTableCell = (dom: DOMUtils, elm: Element) => {
   Optional.from(dom.getParent(elm, 'td,th')).map(SugarElement.fromDom).each(PaddingBr.trimBlockTrailingBr);
 };
 
+const hasInheritableStyles = (styles: string[]): boolean => {
+  // TINY-7326: Figure out what else should go in the nonInheritableStyles list
+  // Elements which have these styles should not get merged as they have non-inherited properties
+  const nonInheritableStyles = [ 'margin', 'padding', 'border', 'background' ];
+  return Arr.forall(styles, (nodeStyle) => Arr.forall(nonInheritableStyles, (nonInheritableStyle) => !Strings.startsWith(nodeStyle, nonInheritableStyle)));
+};
+
+// Remove children nodes that are exactly the same as a parent node - name, attributes, styles
 const reduceInlineTextElements = (editor: Editor, merge: boolean) => {
   const textInlineElements = editor.schema.getTextInlineElements();
   const dom = editor.dom;
+  // TINY-7326: Figure out what else should be added
+  const shorthandCssProps = [ 'font', 'text-decoration', 'text-emphasis' ];
 
   if (merge) {
-    const root = editor.getBody(), elementUtils = ElementUtils(dom);
+    const root = editor.getBody();
+    const elementUtils = ElementUtils(dom);
 
     Tools.each(dom.select('*[data-mce-fragment]'), (node) => {
-      for (let testNode = node.parentNode; testNode && testNode !== root; testNode = testNode.parentNode) {
-        if (textInlineElements[node.nodeName.toLowerCase()] && elementUtils.compare(testNode, node)) {
-          dom.remove(node, true);
+      const nodeStyleProps = Obj.keys(dom.parseStyle(dom.getAttrib(node, 'style')));
+      if (Type.isNonNullable(textInlineElements[node.nodeName.toLowerCase()]) && hasInheritableStyles(nodeStyleProps)) {
+        for (let parentNode = node.parentNode; Type.isNonNullable(parentNode) && parentNode !== root; parentNode = parentNode.parentNode) {
```

Comment:
This is a **very** long function now. Can it be split up into smaller functions?

Reply:
Have split up more now so hopefully it is more readable now